save-agent/src/linuxX64Main/kotlin/com/saveourtool/save/agent/Requests.kt

```diff
@@ -0,0 +1,110 @@
+import com.saveourtool.save.domain.FileKey
+
+import io.ktor.client.*
+import io.ktor.client.call.*
+import io.ktor.client.request.*
+import io.ktor.client.statement.*
+import io.ktor.http.*
+import okio.Path
+import okio.Path.Companion.toPath
+
+/**
+ * Download test source snapshots for execution [executionId] into [target]
+ *
+ * @param config
+ * @param target
+ * @param executionId
+ * @return result
+ */
+internal suspend fun SaveAgent.downloadTestResources(config: BackendConfig, target: Path, executionId: String): Result<Unit> = runCatching {
+    val result = httpClient.downloadTestResources(config, executionId)
+    if (updateStateBasedOnBackendResponse(result)) {
+        return@runCatching
+    }
+
+    val bytes = result.getOrThrow()
+        .readByteArrayOrThrowIfEmpty {
+            error("Not found any tests for execution $executionId")
+        }
+    val pathToArchive = "archive.zip".toPath()
+    logDebugCustom("Writing downloaded archive of size ${bytes.size} into $pathToArchive")
+    bytes.writeToFile(pathToArchive)
+    fs.createDirectories(target, mustCreate = false)
+    pathToArchive.extractZipTo(target)
+    fs.delete(pathToArchive, mustExist = true)
+    logDebugCustom("Extracted archive into $target and deleted $pathToArchive")
+}
+
+/**
+ * Download additional resources from [additionalResourcesAsString] into [targetDirectory]
+ *
+ * @param baseUrl
+ * @param targetDirectory
+ * @param additionalResourcesAsString
+ * @param executionId
+ * @return result
+ */
+internal suspend fun SaveAgent.downloadAdditionalResources(
+    baseUrl: String,
+    targetDirectory: Path,
+    additionalResourcesAsString: String,
+    executionId: String,
+) = runCatching {
+    FileKey.parseList(additionalResourcesAsString)
+        .map { fileKey ->
+            val result = httpClient.downloadFile(
+                "$baseUrl/internal/files/download?executionId=$executionId",
+                fileKey
+            )
+            if (updateStateBasedOnBackendResponse(result)) {
+                return@runCatching
+            }
+
+            val fileContentBytes = result.getOrThrow()
+                .readByteArrayOrThrowIfEmpty {
+                    error("Couldn't download file $fileKey: content is empty")
+                }
+            val targetFile = targetDirectory / fileKey.name
+            fileContentBytes.writeToFile(targetFile)
+            fileKey to targetFile
+        }
+        .onEach { (fileKey, pathToFile) ->
+            pathToFile.markAsExecutable()
+            logDebugCustom(
+                "Downloaded $fileKey into ${fs.canonicalize(pathToFile)}"
+            )
+        }
+        .map { (_, pathToFile) ->
+            unzipIfRequired(pathToFile)
+        }
+        .ifEmpty {
+            logWarn("Not found any additional files for execution \$id")
+            emptyList()
+        }
+}
+
+private suspend fun HttpClient.downloadTestResources(config: BackendConfig, executionId: String) = download(
```

Comment:
PR looks good, but the only thing I do not understand - is how do we plan to merge this logic with the "S3 logic" proposed by @nulls . We will need to download resources from some shared and mounted storage. We will need to decide - who will provide us data. 

If we will make backend a central app that will communicate with S3 or all micro services will do it.

Reply:
How it is currently: FS -> backend -> orchestrator -> agent
How it will be after this PR: FS -> backend -> agent
After we switch from FS to S3 we can either continue like this or teach agent to download from S3 directly, and backend would supply a link to required resources